crypto/verification.go

```diff
@@ -0,0 +1,596 @@
+	mach.Log.Debug("Received verification request from %v", content.FromDevice)
+	otherDevice, err := mach.GetOrFetchDevice(userID, content.FromDevice)
+	if err != nil {
+		mach.Log.Error("Could not find device %v of user %v", content.FromDevice, userID)
+		return
+	}
+	// check if SAS is in supported methods
+	supportsSAS := false
+	for _, method := range content.Methods {
+		if method == event.VerificationMethodSAS {
+			supportsSAS = true
+			break
+		}
+	}
+	if !supportsSAS {
+		mach.Log.Warn("Canceling verification transaction %v as SAS is not supported", content.TransactionID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Only SAS method is supported", event.VerificationCancelUnknownMethod)
+		return
+	}
+	if mach.AcceptVerificationFrom(otherDevice) {
+		mach.Log.Debug("Accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		if err := mach.NewSASVerificationWith(otherDevice, content.TransactionID, 10*time.Minute, true); err != nil {
+			mach.Log.Error("Error accepting SAS verification request: %v", err)
+		}
+	} else {
+		mach.Log.Debug("Not accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Not accepted by user", event.VerificationCancelByUser)
+	}
+}
+
+// NewSASVerificationWith starts the SAS verification process with another device.
+// If the transaction ID is empty, a new one is generated.
+func (mach *OlmMachine) NewSASVerificationWith(device *DeviceIdentity, transactionID string, timeout time.Duration, useEmojiSAS bool) error {
+	if transactionID == "" {
+		transactionID = strconv.Itoa(rand.Int())
+	}
+	mach.Log.Debug("Starting new verification transaction %v with device %v of user %v", transactionID, device.DeviceID, device.UserID)
+
+	verState := &verificationState{
+		sas:                 olm.NewSAS(),
+		otherDevice:         device,
+		supportEmojiSAS:     useEmojiSAS,
+		initiatedByUs:       true,
+		verificationStarted: false,
+		keyReceived:         false,
+		sasMatched:          make(chan bool, 1),
+	}
+
+	startEvent, err := mach.StartSASVerification(device.UserID, device.DeviceID, transactionID, useEmojiSAS)
+	if err != nil {
+		return err
+	}
+
+	json, err := json.Marshal(startEvent)
+	if err != nil {
+		return err
+	}
+	canonical, err := canonicaljson.CanonicalJSON(json)
+	if err != nil {
+		return err
+	}
+
+	verState.startEventCanonical = string(canonical)
+	_, loaded := mach.keyVerificationTransactionState.LoadOrStore(device.UserID.String()+":"+transactionID, verState)
+	if loaded {
+		return errors.New("Transaction already exists")
+	}
+
+	mach.Log.Error("emoji %v", useEmojiSAS)
+
+	go mach.timeoutAfter(device.UserID, transactionID, timeout)
```

Comment:
The timeouts for each step should probably be separated, e.g. if you wait for X-1 seconds for the other person to accept the request, you should have more than 1 second to compare the emoji.

Reply:
Done, whenever the transaction progresses it should now start a new timeout.